tests/test-update-from-main.c

```diff
@@ -0,0 +1,123 @@
+  g_autoptr(EosTestClient) client = NULL;
+  g_auto(CmdAsyncStuff) updater_cmd = CMD_ASYNC_STUFF_CLEARED;
+  g_autoptr(GFile) autoupdater_root = NULL;
+  g_autoptr(EosTestAutoupdater) autoupdater = NULL;
+  g_auto(CmdStuff) reaped = CMD_STUFF_CLEARED;
+  g_autoptr(GPtrArray) cmds = NULL;
+  gboolean has_commit;
+
+  server_root = g_file_get_child (fixture->tmpdir, "main");
+  server = eos_test_server_new_quick (server_root,
+                                      3,
+                                      default_vendor,
+                                      default_product,
+                                      FALSE,
+                                      default_ref,
+                                      0,
+                                      keyid,
+                                      default_ostree_path,
+                                      &error);
+  g_assert_no_error (error);
+  g_assert_cmpuint (server->subservers->len, ==, 1u);
+
+  subserver = g_object_ref (EOS_TEST_SUBSERVER (g_ptr_array_index (server->subservers, 0)));
+  client_root = g_file_get_child (fixture->tmpdir, "client");
+  client = eos_test_client_new (client_root,
+                                default_remote_name,
+                                subserver,
+                                default_ref,
+                                default_vendor,
+                                default_product,
+                                &error);
+  g_assert_no_error (error);
+
+  g_hash_table_insert (subserver->ref_to_commit,
+                       g_strdup (default_ref),
+                       GUINT_TO_POINTER (1));
+  eos_test_subserver_update (subserver,
+                             &error);
+  g_assert_no_error (error);
+
+  eos_test_client_run_updater (client,
+                               DOWNLOAD_MAIN,
+                               &updater_cmd,
+                               &error);
+  g_assert_no_error (error);
+
+  autoupdater_root = g_file_get_child (fixture->tmpdir, "autoupdater");
+  autoupdater = eos_test_autoupdater_new (autoupdater_root,
+                                          UPDATE_STEP_APPLY,
+                                          1,
+                                          TRUE,
+                                          &error);
+  g_assert_no_error (error);
+
+  eos_test_client_reap_updater (client,
+                                &updater_cmd,
+                                &reaped,
+                                &error);
+  g_assert_no_error (error);
+
+  cmds = g_ptr_array_new ();
+  g_ptr_array_add (cmds, &reaped);
+  g_ptr_array_add (cmds, autoupdater->cmd);
+  g_assert_true (cmd_stuff_ensure_all_ok_verbose (cmds));
+
+  eos_test_client_has_commit (client,
+                              default_remote_name,
+                              1,
+                              &has_commit,
+                              &error);
+  g_assert_no_error (error);
+  g_assert_true (has_commit);
+}
```

Comment:
It might be worth spliting these out into several tests, e.g.:
  - created_server_has_single_subserver
  - can_run_server_update
  - updated_server_has_single_commit
  - updater_does_not_have_errors